Maple2.Server.Game/GameServer.cs

```diff
@@ -87,6 +93,54 @@ public GameServer(FieldManager.Factory fieldFactory, PacketRouter<GameSession> r
         return gameEvent;
     }
 
+    public Shop? FindShop(GameSession session, int shopId) {
+        using GameStorage.Request db = gameStorage.Context();
+        if (!shopCache.TryGetValue(shopId, out Shop? shop)) {
+            shop = db.GetShop(shopId);
+            if (shop != null) {
+                shopCache[shopId] = shop;
+            }
+            if (shop?.RestockTime == 0) { // everything else would be player-based shops that would get refreshed
+                if (!shopItemCache.TryGetValue(shopId, out Dictionary<int, ShopItem>? shopItems)) {
+                    shopItems = db.GetShopItems(shopId).ToDictionary(item => item.Id);
+                    shopItemCache[shopId] = shopItems;
+                }
+                foreach ((int shopItemId, ShopItem shopItem) in shopItems) {
+                    Item? item = session.Item.CreateItem(shopItem.ItemId, shopItem.Rarity, shopItem.Quantity);
+                    if (item == null) {
+                        continue;
+                    }
+                    shopItem.Item = item;
+                    shop.Items[shopItem.Id] = shopItem;
+                }
+            } else {
+                return shop;
+            }
+        }
+
+        if (shop.Items.Count == 0 && shopItemCache.TryGetValue(shop.Id, out Dictionary<int, ShopItem>? items)) {
+            foreach ((int shopItemId, ShopItem shopItem) in items) {
+                Item? item = session.Item.CreateItem(shopItem.ItemId, shopItem.Rarity, shopItem.Quantity);
```

Comment:
same as above, i would prefer to not create these items in db until after purchase.